pkg/provision/workspace/automount/git-credentials.go

```diff
@@ -0,0 +1,253 @@
+//
+// Copyright (c) 2019-2021 Red Hat, Inc.
+// This program and the accompanying materials are made
+// available under the terms of the Eclipse Public License 2.0
+// which is available at https://www.eclipse.org/legal/epl-2.0/
+//
+// SPDX-License-Identifier: EPL-2.0
+//
+// Contributors:
+//   Red Hat, Inc. - initial API and implementation
+//
+
+package automount
+
+import (
+	"context"
+	"fmt"
+	"path/filepath"
+	"strings"
+
+	dw "github.com/devfile/api/v2/pkg/apis/workspaces/v1alpha2"
+	"github.com/devfile/devworkspace-operator/apis/controller/v1alpha1"
+	"github.com/devfile/devworkspace-operator/pkg/constants"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	k8sclient "sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+)
+
+const gitCredentialsName = "credentials"
+const gitConfigName = "gitconfig"
+const gitConfigLocation = "/etc/" + gitConfigName
+const credentialTemplate = "[credential]\n\thelper = store --file %s\n"
+
+// getDevWorkspaceGitConfig takes care of mounting git credentials and a gitconfig into a devworkspace.
+//	It does so by:
+//		1. Finding all secrets labeled with "controller.devfile.io/git-credential": "true" and grabbing all their credentials
+//			and condensing them into one string
+//		2. Creating and mounting a gitconfig config map to /etc/gitconfig that points to where the credentials are stored
+//		3. Creating and mounting a credentials secret to mountpath/credentials that stores the users git credentials
+func getDevWorkspaceGitConfig(devworkspace *dw.DevWorkspace, client k8sclient.Client, scheme *runtime.Scheme) (*v1alpha1.PodAdditions, error) {
+	namespace := devworkspace.GetNamespace()
+	secrets := &corev1.SecretList{}
+	err := client.List(context.TODO(), secrets, k8sclient.InNamespace(namespace), k8sclient.MatchingLabels{
+		constants.DevWorkspaceGitCredentialLabel: "true",
+	})
+	if err != nil {
+		return nil, err
+	}
+	var credentials []string
+	var mountpath string
+	for _, secret := range secrets.Items {
+		credentials = append(credentials, string(secret.Data[gitCredentialsName]))
+		if val, ok := secret.Annotations[constants.DevWorkspaceMountPathAnnotation]; ok {
+			mountpath = val
+		}
+	}
+
+	podAdditions := &v1alpha1.PodAdditions{}
+	if len(credentials) > 0 {
+		gitCredsName := devworkspace.Status.DevWorkspaceId + "-" + gitConfigName
```

Comment:
I'm concerned that will effectively take a fixed number of secrets and result in a variable number of secrets based on the number of workspaces -- this could cause issues if a quota is in place. 

Since all workspaces in a namespace should be fine sharing the combined credentials secret, we should be okay defining a fixed name (that's unlikely to collide with others, e.g. `devworkspace-merged-git-credentials`, if that fits in the max length) and using that.

Created https://github.com/devfile/devworkspace-operator/issues/580 for follow-up

(Sorry to bump an already-merged PR, just got back from PTO)

Reply:
> effectively take a fixed number of secrets

I just wonder where this requirement is really coming from.
Like I understand limiting deployment/pods, they are heavy...
But configmap is like 1Mb at maximum right? Then we somebody should limit you to use only 10,20... )
But good point. We should save objects if it's possible and does not hurt anyone.